packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +1246,1035 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
     ];
   }
 }
+
+
+// * OverlayPortal Implementation
+//  OverlayPortal is inspired by the
+//  [flutter_portal](https://pub.dev/packages/flutter_portal) package.
+//
+// ** RenderObject hierarchy
+// The widget works by inserting its overlay child's render subtree directly
+// under [Overlay]'s render object (_RenderTheatre).
+// https://user-images.githubusercontent.com/31859944/171971838-62ed3975-4b5d-4733-a9c9-f79e263b8fcc.jpg
+//
+// To ensure the overlay child render subtree does not do layout twice, the
+// subtree must only perform layout after both its _RenderTheatre and the
+// [OverlayPortal]'s render object (_RenderLayoutSurrogateProxyBox) have
+// finished layout. This is handled by _RenderDeferredLayoutBox.
+//
+// ** Z-Index of an overlay child
+// [_OverlayLocation] is a (currently private) interface that allows an
+// [OverlayPortal] to insert its overlay child into a specific [Overlay], as
+// well as specifying the paint order between the overlay child and other
+// children of the _RenderTheatre.
+//
+// Since [OverlayPortal] is only allowed to target ancestor [Overlay]s
+// (_RenderTheatre must finish doing layout before _RenderDeferredLayoutBox),
+// the _RenderTheatre should typically be acquired using an [InheritedWidget]
+// (currently, _RenderTheatreMarker) in case the [OverlayPortal] gets
+// reparented.
+
+/// A class to show, hide and bring to top an [OverlayPortal]'s overlay child
+/// in the target [Overlay].
+///
+/// A [OverlayPortalController] can only be given to at most one [OverlayPortal]
+/// at a time. When an [OverlayPortalController] is moved from one
+/// [OverlayPortal] to another, its [isShowing] state does not carry over.
+///
+/// [OverlayPortalController.show] and [OverlayPortalController.hide] can be
+/// called even before the controller is assigned to any [OverlayPortal], but
+/// they typically should not be called while the widget tree is being rebuilt.
+class OverlayPortalController {
+  /// Creates an [OverlayPortalController], optionally with a String identifier
+  /// `debugLabel`.
+  OverlayPortalController({ String? debugLabel }) : _debugLabel = debugLabel;
+
+  _OverlayPortalState? _attachTarget;
+  int? _showTimestamp;
+  final String? _debugLabel;
+
+  static int _wallTime = kIsWeb
+    ? -9007199254740992 // -2^53
+    : -1 << 63;
+
+  // Returns a unique and monotonically increasing timestamp.
+  int _now() {
+    final int now = _wallTime += 1;
+    assert(_showTimestamp == null || _showTimestamp! < now);
+    assert(_attachTarget?._showTimestamp == null || _attachTarget!._showTimestamp! < now);
+    return now;
+  }
+
+  /// Show the overlay child of the [OverlayPortal] this controller is attached
+  /// to, at the top of the target [Overlay].
+  ///
+  /// When there are more than one [OverlayPortal]s that target the same
+  /// [Overlay], the overlay child of the last [OverlayPortal] to have called
+  /// [show] appears at the top level, unobstructed.
```

Comment:
This reminds me of @antholeole's loupe PR, where he included a [below](https://github.com/flutter/flutter/blob/7f17098f97f5b2642ca405a400df93cd01079e40/packages/flutter/lib/src/widgets/magnifier.dart#L243) parameter in order to allow `show` calls to have control over the paint order.  Would something like that be useful here to allow ordering the z-index?

Reply:
Actually, another way to look at this problem.  We likely want to port the loupe to OverlayPortal when this is merged.  Will that be possible without a `below` parameter, or what would that look like if so?

Reply:
I now see _OverlayLocation, looks like it's a kind of linked list, which sounds powerful, but it's private.